### ISSUE TITLE
[with solution] Current implementation to find best learning rate is highly unstable

### ISSUE THREAD

**Issue by SirJohnFranklin** [manujosephv/pytorch_tabular#237]
**Describe the bug**
Current implementation to find best learning rate is highly unstable if the curve is not perfectly smooth or interval not chosen correctly.

**To Reproduce**
1. Use a tabular model's tune function to receive best learning rate:
```
result = tabular_model.trainer.tune(tabular_model.model, train_loader, valid_loader, lr_find_kwargs=lr_find_kwargs)
```
2. Get result of current implementation: 
```
pt_suggestion_lr = result['lr_find'].suggestion()  # also initializes the _opimal_idx attribute
pt_suggestion_pos = result['lr_find']._optimal_idx
```
3. Use proposed function to get a proper best learning rate returned, where lr_find_fit is defined below:
```
lr_find_results = result['lr_find'].results
bestlr, yfit = WDeepModelTrainer.lr_find_fit(lr_find_results, skip_first=skip_first, skip_last=skip_last)
bestlr = float(bestlr)
```
with proposed function:
```
@staticmethod
def lr_find_fit(lr_find_results, skip_first=20, skip_last=1):
    def stepfunc(x, A=1, mu=0, sig=1, offset=0):
        arg = (mu - x) / (np.sqrt(2) * sig)
        return A / 2. * (1 + erf(arg)) + offset

    p0 = (np.max(lr_find_results['loss']), -2.5, .2, np.mean(lr_find_results['loss']))

    lrs = np.array(lr_find_results['lr'])[skip_first:-skip_last]
    losses = np.array(lr_find_results['loss'])[skip_first:-skip_last]  # get loss values and their corresponding gradients, and get lr values
    losses = np.clip(losses, -1e9, losses[0])  # clip to last loss

    try:
        params, cv = scipy.optimize.curve_fit(stepfunc, np.log10(lrs), losses, p0)
        yfit = stepfunc(np.log10(lrs), *params)
        bestlr = 10 ** params[1]
        return bestlr, yfit
    except Exception as e:
        logger.exception(f"Could not fit step function to loss curve because of {e}")
        return 1e-3, np.zeros_like(losses)
```

**Expected behavior**
Algorithm to find best learning rate should be stable as shown in the pictures. Sometimes the results match quite well, however with the proposed function the best learning rate is found at (almost) all times even though the initial fit parameters are kept constant in all images below but could be used to tune stability. 

**Screenshots**
[img0](https://ibb.co/6NC0Mbr)
[img1](https://ibb.co/x3Gx9h5)
[img2](https://ibb.co/xq5sSXd)
[img3](https://ibb.co/jTZ7tx7)
[img4](https://ibb.co/TkjbjcP)
[img5](https://ibb.co/C6qXrWm)
[img6](https://ibb.co/7tqWk5V)

**Desktop (please complete the following information):**
Current pull of the code working in python 3.10.

**Additional context**
Any suggestions how the algorithm can be improved more are welcome. Everybody is free to include this addition in their pull request.

Cheers, you guys made a great package!

**Comment by manujosephv** [comment]
Thanks @SirJohnFranklin .

And, I think I see your point. You intend to modify the optimal lr find function with the one you suggested, am I right?

PyTorch Tabular just uses the functionality provided by Pytorch Lightning for the optimal LR find. If you want to raise an issue/PR with PyTorch Lightning, it would be great.

**Comment by manujosephv** [comment]
@SirJohnFranklin  If you have raised an issue in PyTorch Lightning, can I close this?
